Blaze Optimization: Sync Blaze campaigns on campaign list

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of #10933 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds a new view for syncing and displaying Blaze campaigns for a site. Highlights:
- Campaigns are sorted by ID. This should be sufficient for listing items by created date DESC, but I will add an update to the data model to include created/end date for campaigns in a subsequent PR.
- All features for the list have been added:
  - Pagination
  - Pull-to-refresh
  - Empty state and loading state. There is no design for these, so I improvised them.
  - Campaign selection
  - Campaign creation CTA handling 
- The only entry point for now is from the Menu tab. I added a local feature flag to hide the WIP from production. ~~Ann has confirmed in a recent comment (pecCkj-13f-p2#comment-572) that we should show campaign creation if no campaign is created - this will be updated in a subsequent PR.~~
- Also in the above comment, Ann mentioned displaying spent budget while I'm displaying total budget currently. That will also need to be updated later.
- Due to the size of this PR, I'll add unit tests in a subsequent PR.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
- Log in as an admin to a site that is eligible for Blaze.
- Navigate to the Menu tab.
- Select Blaze entry point.
- If your site has existing campaigns, they should be listed with the latest items first. Tapping on any item should show a web view with relevant details.
- If your site doesn't have any campaign, an empty view should be displayed.
- Tap the Create CTA, a new web view should be displayed for creating a new campaign.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/woocommerce/woocommerce-ios/assets/5533851/7bb3ecb2-586d-4979-8e73-e4a44e3cf5ab

Empty state:

<img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/7589b8e1-0d5a-40e9-8d7c-920f17d9ad2d" width=320 />


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

itsmeichigo committed:
Add files for campaign list view and view model

itsmeichigo committed:
Update UI for BlazeCampaignListView

itsmeichigo committed:
Prepare for syncing blaze campaigns

itsmeichigo committed:
Implement syncing Blaze campaigns

itsmeichigo committed:
Fix endpoint path for fetching blaze campaigns

itsmeichigo committed:
Update tests for BlazeRemote

itsmeichigo committed:
Remove NavigationView for blaze campaign list

itsmeichigo committed:
Add a new case for BlazeSource

itsmeichigo committed:
Handle creating campaign from campaign list

itsmeichigo committed:
Fix warnings in BlazeRemoteTests

itsmeichigo committed:
Update layout of the campaign list based on sync state

itsmeichigo committed:
Update entry point from hub menu

itsmeichigo committed:
Fix issue with empty state

itsmeichigo committed:
Put the campaign list under a new feature flag

itsmeichigo committed:
Send site to the campaign list view to handle CTAs

itsmeichigo committed:
Update wording for the empty state

itsmeichigo committed:
Fix typo on comment for campaign list source

itsmeichigo committed:
Handle showing campaign details

itsmeichigo committed:
Move URL extension to a separate file

itsmeichigo committed:
Reload campaign list after creation

itsmeichigo commented:
cc @annchichi Please feel free to check out the screencast and screenshot above and let me know if something is not correct. I'm using a large font size in these.

itsmeichigo committed:
Show results immediately after fetching from storage

itsmeichigo committed:
Merge branch 'trunk' into feat/10933-sync-blaze-campaigns

itsmeichigo committed:
Sort campaigns by IDs DESC

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Blaze/BlazeCampaignList/BlazeCampaignListView.swift:
```diff
@@ -0,0 +1,140 @@
+        Group {
+            switch viewModel.syncState {
+            case .results:
+                RefreshableInfiniteScrollList(spacing: Layout.contentSpacing,
```
Nit: What do you think about separating the empty view, loading view and the list content into private variables for readability?

itsmeichigo replied:
Since the view is not too complicated, I prefer keeping everything in one place to avoid having to scroll up and down to view the content. I think we can have different preferred styles, and one can pick whatever they are more comfortable with.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Blaze/BlazeCampaignList/BlazeCampaignListView.swift:
```diff
@@ -0,0 +1,140 @@
+            case .empty:
+                EmptyState(title: Localization.emptyStateTitle,
+                           description: Localization.emptyStateMessage,
+                           image: .emptyProductsImage)
```
Nice work improvising empty state. 🎖️ I recommend asking design for a suitable empty blaze image as it will help promote Blaze better. (If you haven't already asked 😅)

itsmeichigo replied:
I pinged her again on Figma, will update this screen later in another PR if needed.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Blaze/BlazeCampaignList/BlazeCampaignListView.swift:
```diff
@@ -0,0 +1,140 @@
+    }
+}
+
+private extension BlazeCampaignListView {
```
Super nit: Consider keeping this extension for View-related code and creating a new extension for constants and strings.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/RefreshableInfiniteScrollList.swift:
```diff
@@ -22,22 +25,25 @@ struct RefreshableInfiniteScrollList<Content: View>: View {
     ///
     /// - Parameters:
     ///   - isLoading: Whether the list is loading more content. Used to determine whether to show the infinite scroll indicator.
+    ///   - spacing: Spacing between each item in the list.
```
Super nit: `spacing` is the first parameter, and in the comment, it is the second parameter.

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/RefreshableInfiniteScrollList.swift:
TIL about this view.

selanthiraiyan reviewed (approved):
This works great. 🎉 

I left a few non-blocking nits. 

---

A couple of things I missed in the previous reviews. 

<img src="https://github.com/woocommerce/woocommerce-ios/assets/524475/60bb4c29-e6c7-4cab-a8c0-714e827ac834" width="320"/>

1. The title and the "Impressions" text aren't vertically aligned. 
    - I believe this is because we try to make this work better in landscape mode. I am good if we can't fix this.
2. The title and numbers seem to be "Semi-bold" in the design.

itsmeichigo committed:
Separate BlazeCampaignListView extensions

itsmeichigo committed:
Update comment

itsmeichigo committed:
Fix alignment for stats views

itsmeichigo committed:
Update UI for campaign item

itsmeichigo commented:
Thanks @selanthiraiyan for the reviews! I've addressed the issues in your comments and updated the layout of the campaign items, they should look good now:

![Simulator Screenshot - iPhone 14 Pro - 2023-10-16 at 15 31 39](https://github.com/woocommerce/woocommerce-ios/assets/5533851/48de972c-99f0-4e61-8c5c-1a9a93597f6c)
